### DISCUSSION TITLE
How to get NC variable from 840D

### DISCUSSION THREAD

**Discussion by guiap** [dotnetprojects/DotNetSiemensPLCToolBoxLibrary#170] [General]
Hello, can I get NV Variables from Siemens 840 D?

I achieved to connect on 840 D using snap7 lib, but I need to get this NC variable:

LINK("/Channel/State/feedRateIpoOvr",200,"2 1 3 1 7f# /NC 5 0 11",10)

Thank and Regards.

**Comment by Nick135** [comment]
You need the NC-VAR-Selector to get the NcVar Struct.

Then parse to NC_Var or PLCNckTag and then you can read the Variable.

NC_Var S_feedRateIpoOvr = new NC_Var(0x82, 0x40, 0x3, 0x1, 0x7F, 0x1, 0xF, 0x8);

S_feedRateIpoOvr.GetNckTag(1); //1= ChannelNo

Code snippet
````
using (var myConn = new DotNetSiemensPLCToolBoxLibrary.Communication.PLCConnection("test"))
{
   myConn.Configuration.CpuIP = !string.IsNullOrEmpty(IPAddress) ? IPAddress : "192.168.214.1";
   myConn.Configuration.CpuSlot = plcConnection ? 2 : 4;
   myConn.Connect();
   var r1 = myConn.ReadValue(new NC_Var(0x82, 0x41, 0x3, 0x1, 0x7F, 0x1, 0xF, 0x8));
}
````

**Reply by guiap** [reply]
I wil try this, Thanks.

**Reply by guiap** [reply]
Works fine. thanks Nick135.
Could you help me with more one question?
Can I Get the program name?
I have this link:
LINK("/Channel/ProgramInfo/progName",200,"2 1 3 1 7d# /NC -32 0 11",10)

and I tried this way: 
var progName = myConn.ReadValue(new NC_Var(0x82, 0x41, 0x3, 0x1, 0x7D, 0x1, 0xF, 0x8));
Console.WriteLine($"progName: {progName}");

This output: progName: 1,0691195091498424E+74

**Reply by Nick135** [reply]
SPARP_progName = new NC_Var(0x82, 0x41, 0x3, 0x1, 0x7D, 0x1, **0x13, 0x20**);

You use the wrong data type. Program name is a string (char[32])

**Reply by Nick135** [reply]
0xF = double
0x8 = 8 Byte length

**Reply by guiap** [reply]
Thanks again Nick135, works fine. :)

**Reply by guiap** [reply]
Hello again @Nick135 !

Please, could you explain me, how can I get this values : 

new NC_Var(0x82, 0x41, 0x3, 0x1, 0x7D, 0x1, 0x13, 0x20);

With This line: 

LINK("/Channel/ProgramInfo/progName",200,"2 1 3 1 7d# /NC -32 0 11",10)


I managed to get some variables to work.

I can get others variables, but some variables I can't get, this for example: 

LINK("/Channel/State/progStatus",200,"2 1 13 1 7f# /NC 2 0 11",10)

new NC_Var(0x82,0x41,0x13,0x1,0x7F,0x1,0xF,0x8);

but doesn't works.

Other Example it's this:
LINK("/Channel/MachineAxis/actToolBasePos",200,"2 1 2 1 74# /NC 5 0 11",10)

This is a Vector, like [1], [2] , etc...

There is a reference table that i can convert this "LINK("/Channel/ProgramInfo/progName",200,"2 1 3 1 7d# /NC -32 0 11",10)" to new "NC_Var(0x82, 0x41, 0x3, 0x1, 0x7D, 0x1, 0x13, 0x20)" ?

thanks and regards.

**Reply by Nick135** [reply]
@guiap 
Where did you get that LINK(...)?

I told you to use the "NC-VAR-Selector" to get the Addresses.

**Reply by Nick135** [reply]
LINK("/Channel/State/progStatus",200,"2 1 13 1 7f# /NC 2 0 11",10) 
new NC_Var(0x82, 0x40, 0xD, 0x1, 0x7F, 0x1, 0x4, 0x2); // It is a Word and not a Double

LINK("/Channel/MachineAxis/actToolBasePos",200,"2 1 2 1 74# /NC 5 0 11",10)
new NC_Var(0x82, 0x40, 0x2, 0x0, 0x73, 0x1, 0xF, 0x8);
[0] => new NC_Var(0x82, 0x40, 0x2, **0x1**, 0x73, 0x1, 0xF, 0x8);
[1] => new NC_Var(0x82, 0x40, 0x2, **0x2**, 0x73, 0x1, 0xF, 0x8);

**Reply by guiap** [reply]
Hello, I'm using a nc6.nsk file.

I got this in Sinumerik instalation.

**Comment by guiap** [comment]
Hello Could you help-me, please?
I need to known what's the value of Long integer type. I have these types and values 

CHAR	0X13
DOUBLE	0XF
WORD	0X4
boolean	0x1

**Reply by Nick135** [reply]
Int64 = 0x12 (18)

```
switch (this.Typ)
{
	case 1:
		dataType = TagDataType.Bool;
		break;
	case 3:
		dataType = TagDataType.Byte; //eNCK_LE_Int8;
		break;
	case 4:
		dataType = TagDataType.Word;
		break;
	case 5:
		dataType = TagDataType.Int; //eNCK_LE_Int16;
		break;
	case 6:
		dataType = TagDataType.Dword; //eNCK_LE_Uint32;
		break;
	case 7:
		dataType = TagDataType.Dint; //eNCK_LE_Int32;
		break;
	case 8:
		dataType = TagDataType.Float; //eNCK_LE_Float32;
		break;
	case 14:
		dataType = TagDataType.DateTime;
		break;
	case 15:
		dataType = TagDataType.LReal; //eNCK_LE_Float64;
		break;
	case 18:
		dataType = TagDataType.LInt; //eNCK_LE_Int64;
		break;
	case 19:
		dataType = TagDataType.CharArray; //eNCK_LE_String;
		_ArraySize = this.Laenge;
		break;
	default:
		throw new Exception(string.Format("Unknown Type: {0}", this.Typ));
}

**Comment by annesanz** [comment]
Hello,
I tried defining NC_Var objects and it works fine, I can read information from the CNC, using NC-Var Selector to address the items.
![image](https://user-images.githubusercontent.com/10549406/215713842-8faaab67-2b49-4da5-a50d-f55226aaf4e1.png)

The items are read one by one.
However, I would like to read multiple items together, in a single ReadVar call, to avoid overloading the PLC. Is there any function from this project I can use?
Thank you.

**Reply by Nick135** [reply]
Try something like this:
```
using (var myConn = new DotNetSiemensPLCToolBoxLibrary.Communication.PLCConnection("test"))
{
   myConn.Configuration.CpuIP = !string.IsNullOrEmpty(IPAddress) ? IPAddress : "192.168.214.1";
   myConn.Configuration.CpuSlot = plcConnection ? 2 : 4;
   myConn.Connect();
   var tags = new List<PlcNckTag>();
   tags.add(new PlcNckTag(new NC_Var(0x82, 0x41, 0x3, 0x1, 0x7F, 0x1, 0xF, 0x8));
   tags.add(new PlcNckTag(new NC_Var(0x82, 0x41, 0x3, 0x1, 0x7F, 0x1, 0xF, 0x8));
   myConn.ReadValues(tags);
   var ret1 = tags[0].Value;
   var ret2 = tags[1].Value;
}
```

**Comment by rajaraomv** [comment]
Hello
I donot have NC Var selector. please help on where to download  NC-Var Selector from?
